README.md

```diff
@@ -701,6 +734,100 @@ The set is represented in the configuration as a comma separated list.
 ```
 
 
+### Queries
+
+It is sometimes necessary to share code and data between *lbuild* modules,
+which can be difficult when they are split across files and repositories.
+Queries allow you to share functions and computed properties with other modules
+using the global name resolution system.
+
+```python
+def prepare(module, options):
+    # Add queries to module
+    query = Query(...)
+    module.add_query(query)
+
+def build(env):
+    exists = env.has_query(":module:query")
+    # Access query value or use default if query doesn't exist
+    query = env.query(":module:query", default=lambda arg: arg)
+    # if required call the query with arguments
+    results = query(args)
+```
+
+**Note that queries must be stateless, since module build order is not
+guaranteed. You must enforce this property yourself.**
+
+
+#### Query
+
+This wraps any callable object into a query. By default the name is taken from
+the object's name, however, you may overwrite this.
+Note that when using a lambda function, you must provide a name.
+The description is taken from the objects docstring.
+
+```python
+def shared_function(args):
+    """
+    Describe what this query does.
+
+    :param args: what does it need?
+    :returns: what does it return?
+    """
+    return args
+
+query = Query(function=shared_function)
+query = Query(name="different_name",
+              function=shared_function)
+```
+
+
+#### CachedQuery
+
+This is a convenience wrapper for using `@lru_cache`. See the `functools`
+documentation for details.
+
+```python
+query = CachedQuery(name="cached_function",
+                    function=shared_function,
+                    maxsize=128, typed=False)
+
+# convenience wrapper for this
+@functools.lru_cache(maxsize=128, typed=False)
+def shared_function(args):
+    return args
+query = Query(name="cached_function",
+              function=shared_function)
+```
+
+
+#### PropertyQuery
+
+This query's result is computed only once on demand and then cached.
```

Comment:
Hm, I guess I was thinking of the (Cached)Query more like a jinja filter where you share a "modifier" (technically a pure function).
The property is just for sharing data (which should probably be deepcopied? Currently any calling module can modify it.).

The issue is that the `@lru_cache` decorator for `CachedQuery` won't work correctly when passing the environment, since the `env` is unique for every module.
But this is a general problem, I cannot stop the user from passing `env` manually.
Maybe we should remove the `CachedQuery` completely? You can build it manually after all.

Reply:
I've removed `CachedQuery` completely. You can still decorate your function manually if you really want to and if you know what you're doing.